refactor: optimize Arena Tags save data & Wish persists in save

DayKev opened the pull request:
## What are the changes the user will see?
- Wish now correctly persists across save/load.
- Arena tags now display in the arena flyout again.

## Why am I making these changes?
- Part of #1320

## What are the changes from a developer perspective?
### Text from PKR 6110
- Added missing implementation of `loadTag` to `WishTag`.
- Added many new types and type helpers for the purposes of type-safe parsing of arena data.
- Added a lengthy description at the top of `arena-tag.ts` that discusses changes
- Added a new abstract class, `SerializableArenaTag` to mark arena tags that should persist in save data.
- Simplified the body of `ArenaData`'s constructor to be more readable.
  - When creating an Arena, tags that are not Serializable are now ignored when constructing this data.

### Additional Changes
- Max layers for entry hazards now defined as a getter on the tag class.
- Fixed missing usages of `enumValueToKey` for `ArenaTagType`.

## How to test the changes?
Use various moves that create `ArenaTag`s (Wish, Reflect, hazard moves like Stealth Rocks, etc), then save and reload the save.

## Checklist
- [x] Otherwise: **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`pnpm test:silent`)

DayKev committed:
refactor: Arena Tags no longer store unnecessary data to saves

No more `any` in `ArenaTag#loadTag` or `ArenaData`'s constructor

Fix Wish not persisting across saves

Fix arena tags not displaying in the arena flyout

innerthunder commented on src/data/arena-tags/delayed-attack-tag.ts:
```diff
@@ -41,12 +52,12 @@ export class DelayedAttackTag extends ArenaTag {
    */
   public addAttack(source: Pokemon, moveId: MoveId, targetIndex: FieldBattlerIndex): void {
-    this.delayedAttacks.push({ sourceId: source.id, moveId: moveId, targetIndex, turnCount: 3 });
+    (this.delayedAttacks as DelayedAttack[]).push({ sourceId: source.id, moveId: moveId, targetIndex, turnCount: 3 });
```
```suggestion
    (this.delayedAttacks as DelayedAttack[]).push({ sourceId: source.id, moveId, targetIndex, turnCount: 3 });
```

innerthunder commented on src/data/arena-tags/entry-hazard-tag.ts:
```diff
@@ -1,33 +1,31 @@
+  public abstract override readonly tagType: EntryHazardTagType;
+  public readonly layers: number = 1;
+
+  public abstract get maxLayers(): number;
```
This could use a doc like `WeakenMoveScreenTag` explaining why this is a getter and not just `public readonly`

innerthunder commented on src/data/arena-tags/toxic-spikes-tag.ts:
```diff
@@ -16,11 +16,17 @@ import i18next from "i18next";
-  constructor(sourceId: number, side: ArenaTagSide) {
-    super(ArenaTagType.TOXIC_SPIKES, MoveId.TOXIC_SPIKES, sourceId, side, 2);
-    this.neutralized = false;
+  #neutralized: boolean;
```
Why `#` instead of `private`?

DayKev replied:
This field doesn't need to be serialized because it can only ever be `true` right before the tag is removed.

A field that is `private` or `protected` will still be serialized, however TypeScript won't be able to see it and so it will cause type errors, which is why fields that need to be serialized but are normally supposed to be `private` need the `readonly` stuff so they can't be modified outside the class. Basically, working around TypeScript limitations.

innerthunder commented on src/data/arena-tags/wish-tag.ts:
```diff
@@ -1,46 +1,69 @@
-  private triggerMessage: string;
-  private healHp: number;
+export class WishTag extends SerializableArenaTag {
+  // The following fields are meant to be inwardly mutable, but outwardly immutable.
+  public readonly battlerIndex: BattlerIndex;
+  public readonly healHp: number;
+  public readonly sourceName: string;
+  // End inwardly mutable fields
```
Out of this PR's scope, but this doesn't support Wish being used in both slots on the same side in a double battle. It might be worth adding an `edgeCase` for now...

DayKev replied:
Added. There's a PR to rework Wish and Future Sight at PKR that fixes the issue that should probably be ported at some point. https://github.com/pagefaultgames/pokerogue/pull/5862

innerthunder commented on src/system/game-data.ts:
```diff
@@ -891,7 +892,9 @@ export class GameData {
-                globalScene.arena.eventTarget.dispatchEvent(new TagAddedEvent(tag.tagType, tag.side, tag.turnCount));
+                globalScene.arena.eventTarget.dispatchEvent(
+                  new TagAddedEvent(tag.tagType, tag.side, tag.turnCount, 0, 0),
+                );
```
We can do this to condense the `if/else` here:

```ts
for (const tag of globalScene.arena.tags) {
  const { tagType, side, turnCount } = tag;
  const { layers, maxLayers } = tag instanceof EntryHazardTag ? tag : {};
  globalScene.arena.eventTarget.dispatchEvent(
    new TagAddedEvent(tagType, side, turnCount, layers ?? 0, maxLayers ?? 0);
  ); 
}

DayKev committed:
Apply review suggestions

- Remove redundant `moveId: moveId` from `delayed-attack-tag.ts`

- Add docs to `EntryHazardTag#maxLayers`

- Simplify logic for emitting `TagAddedEvent`s in `GameData`

DayKev committed:
Add `.edgeCase()` note to Wish

DayKev committed:
Merge branch 'beta' into refactor/arena-tag-save-data

DayKev committed:
Merge branch 'beta' into refactor/arena-tag-save-data

DayKev committed:
Merge branch 'beta' into refactor/arena-tag-save-data

DayKev committed:
Merge branch 'beta' into refactor/arena-tag-save-data

DayKev committed:
Fix merge issue

DayKev committed:
Merge branch 'beta' into refactor/arena-tag-save-data

DayKev committed:
Merge branch 'beta' into refactor/arena-tag-save-data

DayKev committed:
Merge branch 'beta' into refactor/arena-tag-save-data

DayKev committed:
Merge branch 'beta' into refactor/arena-tag-save-data

DayKev committed:
Apply Biome

DayKev committed:
Resolve merge issue

DayKev committed:
Merge branch 'beta' into refactor/arena-tag-save-data

DayKev committed:
Merge branch 'beta' into refactor/arena-tag-save-data

DayKev committed:
Merge branch 'beta' into refactor/arena-tag-save-data

DayKev committed:
Apply Biome

DayKev committed:
Merge branch 'beta' into refactor/arena-tag-save-data

DayKev committed:
Merge branch 'beta' into refactor/arena-tag-save-data

DayKev committed:
Merge branch 'beta' into refactor/arena-tag-save-data

DayKev committed:
Merge branch 'beta' into refactor/arena-tag-save-data

DayKev committed:
Merge branch 'beta' into refactor/arena-tag-save-data

DayKev committed:
Fix merge issues

DayKev committed:
Fix `@module` tag spec in `tsdoc.json`

DayKev committed:
Merge branch 'beta' into refactor/arena-tag-save-data
